components/script/script_thread.rs

```diff
@@ -832,11 +828,19 @@ impl ScriptThread {
                         new_layout_info)) => {
                     self.profile_event(ScriptThreadEventCategory::AttachLayout, || {
                         // If this is an about:blank load, it must share the creator's origin.
-                        let origin = new_layout_info.load_data.creator_pipeline_id.and_then(|creator_pipeline_id| {
-                            self.documents.borrow().find_document(creator_pipeline_id).map(|doc| {
-                                doc.origin().clone()
-                            })
-                        }).unwrap_or(MutableOrigin::new(new_layout_info.load_data.url.origin()));
+                        let origin = if new_layout_info.load_data.url.as_str() != "about:blank" {
+                            MutableOrigin::new(new_layout_info.load_data.url.origin())
+                        } else if let Some(parent) = new_layout_info.parent_info
+                                .and_then(|(pipeline_id, _)| self.documents.borrow()
+                                .find_document(pipeline_id)) {
+                            parent.origin().clone()
+                        } else if let Some(creator) = new_layout_info.load_data.creator_pipeline_id
+                                .and_then(|pipeline_id| self.documents.borrow()
+                                .find_document(pipeline_id)) {
+                            creator.origin().clone()
+                        } else {
+                            MutableOrigin::new(ImmutableOrigin::new_opaque())
+                        };
 
```

Comment:
This looks good. We should probably put comments here and in constellation saying that these two bits of code need to agree with each other.